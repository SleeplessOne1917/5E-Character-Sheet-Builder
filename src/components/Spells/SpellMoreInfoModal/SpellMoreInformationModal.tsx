import { MutableRefObject, useCallback, useEffect, useRef } from 'react';

import Button from '../../Button/Button';
import LoadingSpinner from '../../LoadingSpinner/LoadingSpinner';
import ModalBackground from '../../ModalBackground/ModalBackground';
import { Spell } from '../../../types/characterSheetBuilderAPI';
import { XCircleIcon } from '@heroicons/react/24/solid';
import classes from './SpellMoreInformationModal.module.css';
import dynamic from 'next/dynamic';
import { useRouter } from 'next/router';

const MarkdownParser = dynamic(
	() => import('../../MarkdownParser/MarkdownParser'),
	{
		loading: () => (
			<div
				style={{
					display: 'flex',
					alignItems: 'center',
					justifyContent: 'center'
				}}
			>
				<LoadingSpinner />
			</div>
		)
	}
);

type SpellMoreInformationModalProps = {
	shouldShowEditAndClasses?: boolean;
	show: boolean;
	spell?: Spell;
	onClose: () => void;
};

const SpellMoreInformationModal = ({
	show,
	spell,
	onClose,
	shouldShowEditAndClasses = false
}: SpellMoreInformationModalProps) => {
	const closeButtonRef = useRef<HTMLButtonElement>();
	const router = useRouter();

	const handleEditClick = useCallback(() => {
		router.push(`/my-stuff/edit/spells/${spell?.id}`);
	}, [router, spell?.id]);

	useEffect(() => {
		if (show) {
			closeButtonRef.current?.focus();
		}
	}, [show]);

	return (
		<ModalBackground testId="spell-more-information-modal" show={show}>
			<div className={classes.modal}>
				<div className={classes['title-container']}>
					<svg className={classes['title-icon']}>
						<use xlinkHref={`/Icons.svg#${spell?.school.id}`} />
					</svg>
					<h3>{spell?.name}</h3>
				</div>
				<div className={classes['close-container']}>
					<Button
						onClick={onClose}
						style={{
							display: 'flex',
							alignItems: 'center'
						}}
						ref={closeButtonRef as MutableRefObject<HTMLButtonElement>}
					>
						<XCircleIcon className={classes['close-icon']} />
						Close
					</Button>
				</div>
				<div className={classes['content-container']}>
					<div className={classes.summary}>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Level</span>:{' '}
							{spell?.level === 0 ? 'Cantrip' : spell?.level}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>
								Casting time
							</span>
							: {spell?.castingTime}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Duration</span>:{' '}
							{spell?.duration}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Range</span>:{' '}
							{spell?.range}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>School</span>:{' '}
							{spell?.school.name}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Components</span>:{' '}
							{spell?.components.join(', ')}
							{spell?.material ? ` (${spell.material})` : ''}
						</div>
						{shouldShowEditAndClasses && (
							<div className={classes['summary-item']}>
								<span className={classes['summary-item-label']}>Classes</span>:{' '}
								{spell?.classes.map(({ name }) => name).join(', ')}
							</div>
						)}
					</div>
					<div className={classes.description}>
						<MarkdownParser input={spell?.description ?? ''} />
						{spell?.atHigherLevels && (
							<MarkdownParser
								input={'**At Higher Levels**: ' + spell.atHigherLevels}
							/>
						)}
					</div>
					<div className={classes.other}>
						{spell?.concentration && (
							<p className={classes['other-info']}>Requires concentration.</p>
						)}
						{spell?.ritual && (
							<p className={classes['other-info']}>Can be cast as a ritual.</p>
						)}
						{spell?.damageType && (
							<div className={classes['damage-display']}>
								<div>
									<span className={classes['summary-item-label']}>
										Damage type
									</span>
									: {spell.damageType?.name}
								</div>
								<svg className={classes['damage-icon']}>
									<use xlinkHref={`/Icons.svg#${spell.damageType?.id}`} />
								</svg>
							</div>
						)}
						{shouldShowEditAndClasses && (
							<Button
								positive
								onClick={handleEditClick}
								style={{ width: '80%' }}
							>
								Edit
							</Button>
						)}
					</div>
				</div>
			</div>
		</ModalBackground>
	);
};

export default SpellMoreInformationModal;
