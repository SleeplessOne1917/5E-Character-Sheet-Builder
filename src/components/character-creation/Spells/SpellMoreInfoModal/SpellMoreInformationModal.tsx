import { useEffect, useRef } from 'react';

import Button from '../../../Button/Button';
import { SrdSpellItem } from '../../../../types/srd';
import { XCircleIcon } from '@heroicons/react/24/solid';
import classes from './SpellMoreInformationModal.module.css';
import ModalBackground from '../../../ModalBackground/ModalBackground';
import MarkdownParser from '../../../MarkdownParser/MarkdownParser';

type SpellMoreInformationModalProps = {
	show: boolean;
	spell?: SrdSpellItem;
	onClose: () => void;
};

const SpellMoreInformationModal = ({
	show,
	spell,
	onClose
}: SpellMoreInformationModalProps) => {
	const closeButtonRef = useRef<HTMLButtonElement>();

	useEffect(() => {
		if (show) {
			closeButtonRef.current?.focus();
		}
	}, [show]);

	return (
		<ModalBackground testId="spell-more-information-modal" show={show}>
			<div className={classes.modal}>
				<div className={classes['title-container']}>
					<svg className={classes['title-icon']}>
						<use xlinkHref={`/Icons.svg#${spell?.school.index}`} />
					</svg>
					<h3>{spell?.name}</h3>
				</div>
				<div className={classes['close-container']}>
					<Button
						onClick={onClose}
						style={{
							display: 'flex',
							alignItems: 'center'
						}}
						ref={closeButtonRef}
					>
						<XCircleIcon className={classes['close-icon']} />
						Close
					</Button>
				</div>
				<div className={classes['content-container']}>
					<div className={classes.summary}>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Level</span>:{' '}
							{spell?.level === 0 ? 'Cantrip' : spell?.level}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>
								Casting time
							</span>
							: {spell?.casting_time}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Duration</span>:{' '}
							{spell?.duration}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Range</span>:{' '}
							{spell?.range}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>School</span>:{' '}
							{spell?.school.name}
						</div>
						<div className={classes['summary-item']}>
							<span className={classes['summary-item-label']}>Components</span>:{' '}
							{spell?.components.join(', ')}
							{spell?.material ? ` (${spell.material})` : ''}
						</div>
					</div>
					<div className={classes.description}>
						<MarkdownParser input={spell?.desc ?? ''} />
						{spell?.higher_level &&
							spell.higher_level.map((hl, index) => (
								<p key={`${spell.index}-higher-level-${index}`}>
									{index === 0 && (
										<span className={classes['summary-item-label']}>
											At Higher Levels:{' '}
										</span>
									)}
									{hl}
								</p>
							))}
					</div>
					<div className={classes.other}>
						{spell?.concentration && (
							<p className={classes['other-info']}>Requires concentration.</p>
						)}
						{spell?.ritual && (
							<p className={classes['other-info']}>Can be cast as a ritual.</p>
						)}
						{spell?.damage && spell.damage.damage_type && (
							<div className={classes['damage-display']}>
								<div>
									<span className={classes['summary-item-label']}>
										Damage type
									</span>
									: {spell.damage.damage_type.name}
								</div>
								<svg className={classes['damage-icon']}>
									<use
										xlinkHref={`/Icons.svg#${spell.damage.damage_type.index}`}
									/>
								</svg>
							</div>
						)}
					</div>
				</div>
			</div>
		</ModalBackground>
	);
};

export default SpellMoreInformationModal;
