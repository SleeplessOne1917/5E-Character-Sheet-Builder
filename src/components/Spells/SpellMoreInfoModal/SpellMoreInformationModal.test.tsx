import * as Router from 'next/navigation';
import * as stories from './SpellMoreInformationModal.stories';

import { render, screen } from '@testing-library/react';

import { composeStories } from '@storybook/react';

const { Default, Loading, Error } = composeStories(stories);

const mockRouter = (pathname: string) => {
	const useRouter = jest.spyOn(Router, 'useRouter');

	const router: Router.NextRouter = {
		push: () => Promise.resolve(true),
		prefetch: () => Promise.resolve(),
		pathname,
		route: '',
		query: {},
		asPath: '',
		basePath: '',
		isLocaleDomain: true,
		reload: jest.fn(),
		replace: jest.fn(),
		back: jest.fn(),
		beforePopState: jest.fn(),
		events: { emit: jest.fn(), on: jest.fn(), off: jest.fn() },
		isFallback: false,
		isPreview: false,
		isReady: true,
		forward: () => {}
	};
	useRouter.mockReturnValue(router);
};

describe('renders correctly', () => {
	beforeEach(() => {
		mockRouter('/my-stuff/spells');
	});

	it('default', () => {
		render(<Default />);

		expect(
			screen.getByTestId('spell-more-information-modal')
		).toMatchSnapshot();
	});

	it('loading', () => {
		render(<Loading />);

		expect(
			screen.getByTestId('spell-more-information-modal')
		).toMatchSnapshot();
	});

	it('error', () => {
		render(<Error />);

		expect(
			screen.getByTestId('spell-more-information-modal')
		).toMatchSnapshot();
	});
});
