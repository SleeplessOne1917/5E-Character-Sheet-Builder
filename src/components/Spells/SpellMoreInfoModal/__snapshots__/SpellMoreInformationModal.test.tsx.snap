// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  class="modal-container"
  data-testid="spell-more-information-modal"
  style="display: flex;"
>
  <div
    class="modal"
  >
    <div
      class="title-container"
    >
      <svg
        class="title-icon"
      >
        <use
          xlink:href="/Icons.svg#illusion"
        />
      </svg>
      <h3>
        Fear
      </h3>
    </div>
    <div
      class="close-container"
    >
      <button
        class="button negative"
        style="font-size: 1.3rem; padding: 1rem; margin: 0.25rem; display: flex; align-items: center;"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="close-icon"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zm-1.72 6.97a.75.75 0 10-1.06 1.06L10.94 12l-1.72 1.72a.75.75 0 101.06 1.06L12 13.06l1.72 1.72a.75.75 0 101.06-1.06L13.06 12l1.72-1.72a.75.75 0 10-1.06-1.06L12 10.94l-1.72-1.72z"
            fill-rule="evenodd"
          />
        </svg>
        Close
      </button>
    </div>
    <div
      class="content-container"
    >
      <div
        class="summary"
      >
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            Level
          </span>
          :
           
          3
        </div>
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            Casting time
          </span>
          : 
          1 action
        </div>
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            Duration
          </span>
          :
           
        </div>
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            Range
          </span>
          :
           
          Self
        </div>
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            School
          </span>
          :
           
          Illusion
        </div>
        <div
          class="summary-item"
        >
          <span
            class="summary-item-label"
          >
            Components
          </span>
          :
           
          V, S, M
           (A white feather or the heart of a hen.)
        </div>
      </div>
      <div
        class="description"
      >
        <div
          style="display: flex; align-items: center; justify-content: center;"
        >
          <div
            class="container"
          >
            <svg
              class="icon"
            >
              <use
                xlink:href="/Icons.svg#spinner"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        class="other"
      >
        <p
          class="other-info"
        >
          Requires concentration.
        </p>
      </div>
    </div>
  </div>
</div>
`;
